### PR TITLE
(feat): convenient helper functions

### PR DESCRIPTION
**Pull request by tareqimbasher**
add get, post, put, patch and delete functions that improve ease of use and readability. before change, making fetch requests other than GET involved explicitly defining options to set HTTP method and body.

I created some unit tests although I don't know how effective they are, and exactly how to write them for a more meaningful assertion.

### CODE REVIEW

**Comment by tareqimbasher** [comment]
Hi guys, besides the above pull request I also typically use an extended version of the `HttpClient` in my projects where calls to the helper functions introduced in the above pull request return a custom `HttpResult` that looks a bit like this:

```
interface IHttpResult {
    success: boolean;
    response: Response | undefined;
    bodyString?: string;
    isJsonBody: boolean;

    bodyAs<TDestType>(ctor: INoParamConstructor<TDestType>): TDestType;
    bodyAsArray<TDestType>(ctor: INoParamConstructor<TDestType>): TDestType[];
}

interface INoParamConstructor<T> {
    new (): T;
}
```

Usage would look like (using TypeScript):

```
class Thing {   // has a parameterless constructor
    name: string;
}

let result = await this.http.get('path/to/resource');
if (result.success) {
    let thing = result.bodyAs(Thing);

    // or
    let things = result.bodyAsArray(Thing);
}
```

I can also give more details on this. The "extended" version of the `HttpClient` I use also is error-safe so if there is an error during the call or if the result itself is "not successful," you can be assured you will always get back a HttpResult with the errors not an exception that you need to handle everytime you make a fetch call. This also helps with the awkward way we need to pass an "error callback" to handle unsuccessful fetch requests.

If you think that would be something helpful to add (which I obviously do) I would be happy to submit a pull request for your review.

File: src/http-client.js [hunk 1]
```diff
@@ -123,6 +123,75 @@ export class HttpClient {
...
+  *
+  * @param input The resource that you wish to fetch. Either a
+  * Request object, or a string containing the URL of the resource.
+  * @param body The body of the request.
+  * @param init An options object containing settings to be applied to
+  * the Request.
+  * @returns A Promise for the Response from the fetch request.
+  */
+  post(input: Request|string, body?: any, init?: RequestInit): Promise<Response> {
+    return this::callFetch(input, body, init, 'post');
```
Comment (EisenbergEffect) [inline src/http-client.js]: As much as I love the bind operator, it's still a bit early for us to use this I think. Would you mind switching this to more standard JS?

File: test/http-client.spec.js [hunk 2]
```diff
@@ -202,6 +202,91 @@ describe('HttpClient', () => {
...
+          expect(fetch).toHaveBeenCalled();
+          done();
+        });
+    });
+  });
+
+  describe('patch', () => {
+    it('sets method to PATCH and body of request and calls fetch', () => {
+      client
+        .put('http://example.com/some/cool/path', JSON.stringify({ test: 'object' }))
```
Comment (EisenbergEffect) [inline test/http-client.spec.js]: Looks like this test isn't actually testing patch.

File: src/http-client.js (hunk 1 above)
Comment (AshleyGrant) [reply src/http-client.js]: Actually, the bind operator is already being used a decent amount in this codebase. Here's an example: https://github.com/aurelia/fetch-client/blob/master/src/http-client.js#L107

Looks like it's used six times in the current http-client.js

Comment (EisenbergEffect) [reply src/http-client.js]: Doh. We should fix that then. I don't even see it on the proposals list anymore...

**Comment by tareqimbasher** [comment]
@EisenbergEffect @AshleyGrant Thank you guys. Sorry was out of town, I will fix issues you pointed out and update soon.

**Comment by tareqimbasher** [comment]
I have updated the unit test to correctly test patch. I left out the fixes requested by @EisenbergEffect first comment (about the bind operator in `this::callFetch`) since as @AshleyGrant mentioned that is already the standard used in the existing implementation.

I would be happy to change all the bind operators into more standard JS. Let me know if you'd like me to submit a pull request for that.

**Pull request merged**